### PR TITLE
middleware: add basic response middleware

### PR DESCRIPTION
**Pull request by joshdekock**
This commit adds a basic response middlware for both announces, and scrapes.

Completes one of the tasks in #145

### CODE REVIEW

File: server/store/middleware/response/response.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		storage := store.MustGetStore()
+		resp.IPv4Peers, resp.IPv6Peers, err = storage.AnnouncePeers(req.InfoHash, req.Left == 0, int(req.NumWant))
+		if err != nil {
+			return ErrCouldntGetPeers
+		}
+		resp.Compact = req.Compact
+		resp.Incomplete = storage.NumLeechers(req.InfoHash)
+		resp.Complete = storage.NumSeeders(req.InfoHash)
+
+		next(cfg, req, resp)
```
Comment (mrd0ll4r) [inline server/store/middleware/response/response.go]: this can return an error

File: server/store/middleware/response/response.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		var scrape chihaya.Scrape
+		storage := store.MustGetStore()
+		for _, infoHash := range req.InfoHashes {
+			resp.Files[infoHash] = chihaya.Scrape{
+				Complete:   storage.GetSeeders(infoHash),
+				Incomplete: storage.GetLeechers(infoHash),
+			}
+		}
+
+		next(cfg, req, resp)
```
Comment (mrd0ll4r) [inline server/store/middleware/response/response.go]: this can also return an error :)

File: server/store/middleware/response/response.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// responseScrapeClient provides a middleware to make a response to an
+// scrape based on the current request.
+func responseScrapeClient(next tracker.ScrapeHandler) tracker.ScrapeHandler {
+	return func(cfg *chihaya.TrackerConfig, req *chihaya.ScrapeRequest, resp *chihaya.ScrapeResponse) (err error) {
+		var scrape chihaya.Scrape
+		storage := store.MustGetStore()
+		for _, infoHash := range req.InfoHashes {
+			resp.Files[infoHash] = chihaya.Scrape{
+				Complete:   storage.GetSeeders(infoHash),
+				Incomplete: storage.GetLeechers(infoHash),
```
Comment (mrd0ll4r) [inline server/store/middleware/response/response.go]: scrapes contain just counts, not actual lists of peers

File: server/store/middleware/response/README.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+## Response Middleware
+
+This package provides the announce middlewares `response` for blacklisting or whitelisting IP addresses and networks for announces.
```
Comment (jzelinskie) [inline server/store/middleware/response/README.md]: What does this have to do with blacklisting and whitelisting?

File: server/store/middleware/response/response.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"github.com/chihaya/chihaya"
+	"github.com/chihaya/chihaya/server/store"
+	"github.com/chihaya/chihaya/tracker"
+)
+
+func init() {
+	tracker.RegisterAnnounceMiddleware("response", responseAnnounceClient)
+	tracker.RegisterScrapeMiddleware("reponse", responseScrapeClient)
```
Comment (jzelinskie) [inline server/store/middleware/response/response.go]: I think the name should also have the prefix "store_response" because it depends on the store server.

File: server/store/middleware/response/README.md (hunk 4 above)
Comment (joshdekock) [inline server/store/middleware/response/README.md]: err, I guess I forgot to change the description

File: server/store/middleware/response/response.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// ErrCouldntGetPeers is returned when something goes wrong with getting
+// peers from the store.
+var ErrCouldntGetPeers = tracker.ClientError("couldn't get peers")
+
+// responseAnnounceClient provides a middleware to make a response to an
+// announce based on the current request.
+func responseAnnounceClient(next tracker.AnnounceHandler) tracker.AnnounceHandler {
+	return func(cfg *chihaya.TrackerConfig, req *chihaya.AnnounceRequest, resp *chihaya.AnnounceResponse) (err error) {
```
Comment (jzelinskie) [inline server/store/middleware/response/response.go]:
```go
func responseAnnounceClient(next tracker.AnnounceHandler) tracker.AnnounceHandler {
	return func(cfg *chihaya.TrackerConfig, req *chihaya.AnnounceRequest, resp *chihaya.AnnounceResponse) (err error) {
		storage := store.MustGetStore()

		resp.MinInterval = cfg.MinAnnounceInterval
		resp.Compact = req.Compact
		resp.Complete = storage.NumSeeders(req.InfoHash)
		resp.Incomplete = storage.NumLeechers(req.InfoHash)
		resp.IPv4Peers, resp.IPv6Peers, err = storage.AnnouncePeers(req.InfoHash, req.Left == 0, int(req.NumWant))
		if err != nil {
			return ErrCouldntGetPeers
		}

		return next(cfg, req, resp)
	}
}
```

Comment (joshdekock) [inline server/store/middleware/response/response.go]: Done.

File: server/store/middleware/response/response.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+func init() {
+	tracker.RegisterAnnounceMiddleware("store_response", responseAnnounceClient)
+	tracker.RegisterScrapeMiddleware("store_response", responseScrapeClient)
+}
+
+// ErrCouldntGetPeers is returned when something goes wrong with getting
+// peers from the store.
+var ErrCouldntGetPeers = tracker.ClientError("couldn't get peers")
```
Comment (jzelinskie) [inline server/store/middleware/response/response.go]: Maybet this should actually be a type instead of a variable. That way we can wrap the actual error we get with this type and not loose any context.

**Comment by jzelinskie** [comment]
LGTM
